CLI panic in processIgnoreChanges

### What happened?

```
panic: runtime error: index out of range [1] with length 1

goroutine 148 [running]:
github.com/pulumi/pulumi/sdk/v3/go/common/resource.PropertyPath.reset({0x140024236c0, 0x2, 0x4}, {{0x102d78520?, 0x140002ff1b8?}}, {{0x102d78520?, 0x1400217f7e8?}}, 0x0, 0x0)
/home/runner/work/pulumi/pulumi/sdk/go/common/resource/properties_path.go:464 +0x870
github.com/pulumi/pulumi/sdk/v3/go/common/resource.PropertyPath.reset({0x140024236b0, 0x3, 0x5}, {{0x10322ba80?, 0x14002260420?}}, {{0x10322ba80?, 0x14001be08a0?}}, 0x0, 0x0)
/home/runner/work/pulumi/pulumi/sdk/go/common/resource/properties_path.go:527 +0x4a0
github.com/pulumi/pulumi/sdk/v3/go/common/resource.PropertyPath.reset({0x140024236a0, 0x4, 0x6}, {{0x10322ba80?, 0x140022603f0?}}, {{0x10322ba80?, 0x14001be0870?}}, 0x0, 0x0)
/home/runner/work/pulumi/pulumi/sdk/go/common/resource/properties_path.go:527 +0x4a0
github.com/pulumi/pulumi/sdk/v3/go/common/resource.PropertyPath.reset({0x14002423690, 0x5, 0x7}, {{0x102d78520?, 0x140002ff368?}}, {{0x102d78520?, 0x1400217f800?}}, 0x0, 0x0)
/home/runner/work/pulumi/pulumi/sdk/go/common/resource/properties_path.go:464 +0x844
github.com/pulumi/pulumi/sdk/v3/go/common/resource.PropertyPath.reset({0x14002423680, 0x6, 0x8}, {{0x10322ba80?, 0x140022603c0?}}, {{0x10322ba80?, 0x14001c00360?}}, 0x0, 0x0)
/home/runner/work/pulumi/pulumi/sdk/go/common/resource/properties_path.go:527 +0x4a0
github.com/pulumi/pulumi/sdk/v3/go/common/resource.PropertyPath.Reset(...)
/home/runner/work/pulumi/pulumi/sdk/go/common/resource/properties_path.go:540
github.com/pulumi/pulumi/pkg/v3/resource/deploy.processIgnoreChanges(0x14001be0840, 0x140022603c0, {0x14001bd8f40, 0x1, 0x10318f400?})
/home/runner/work/pulumi/pulumi/pkg/resource/deploy/step_generator.go:1889 +0x18c
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*stepGenerator).generateSteps(0x140024b0960, {0x12c501960, 0x14001bd9460})
/home/runner/work/pulumi/pulumi/pkg/resource/deploy/step_generator.go:568 +0x860
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*stepGenerator).GenerateSteps(0x140024b0960, {0x12c501960?, 0x14001bd9460?})
/home/runner/work/pulumi/pulumi/pkg/resource/deploy/step_generator.go:281 +0x30
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*deploymentExecutor).handleSingleEvent(0x1400258bd10, {0x103341540, 0x14001bd9460})
/home/runner/work/pulumi/pulumi/pkg/resource/deploy/deployment_executor.go:446 +0x5d8
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*deploymentExecutor).Execute.func3(0x14002497080, 0x1400258bd10, 0x140025a2850, {0x103369568, 0x1400259a910}, {0x103369530, 0x1400258bb60})
/home/runner/work/pulumi/pulumi/pkg/resource/deploy/deployment_executor.go:255 +0x178
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*deploymentExecutor).Execute(0x1400258bd10, {0x103369530, 0x1400258bb60})
/home/runner/work/pulumi/pulumi/pkg/resource/deploy/deployment_executor.go:271 +0x300
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*Deployment).Execute(...)
/home/runner/work/pulumi/pulumi/pkg/resource/deploy/deployment.go:587
github.com/pulumi/pulumi/pkg/v3/engine.(*deployment).run.func1()
/home/runner/work/pulumi/pulumi/pkg/engine/deployment.go:337 +0x88
created by github.com/pulumi/pulumi/pkg/v3/engine.(*deployment).run in goroutine 1
/home/runner/work/pulumi/pulumi/pkg/engine/deployment.go:336 +0x174
```

### Example

Program:

```typescript
import * as aws from "@pulumi/aws";
import * as awsx from "@pulumi/awsx";
import * as pulumi from "@pulumi/pulumi";

const cfg = new pulumi.Config();

const targetGroupCount = cfg.requireNumber("targetGroupCount");

const vpc = new awsx.ec2.Vpc("lb-vpc", {
  cidrBlock: "10.0.0.0/16",
  subnetStrategy: "Auto",
  subnetSpecs: [
    {
      type: "Public",
      name: "ecs-public-subnet",
    },
  ],
  numberOfAvailabilityZones: 2,
  natGateways: {
    strategy: "None"
  }
});

const lbSecurityGroup = new aws.ec2.SecurityGroup("lbSg", {
    vpcId: vpc.vpcId,
    ingress: [{
        protocol: "tcp",
        fromPort: 80,
        toPort: 80,
        cidrBlocks: ["0.0.0.0/0"]
    },
],
    egress: [{
        protocol: "-1",
        fromPort: 0,
        toPort: 0,
        cidrBlocks: ["0.0.0.0/0"]
    }]
});

const lb = new aws.lb.LoadBalancer("lb", {
    securityGroups: [lbSecurityGroup.id],
    subnets: vpc.publicSubnetIds,
});

const tg = new aws.lb.TargetGroup("tg", {
    port: 8080,
    protocol: "HTTP",
    targetType: "ip",
    vpcId: vpc.vpcId,
    deregistrationDelay: 5
}, {dependsOn: [lb]});

const tg2 = new aws.lb.TargetGroup("tg2", {
    port: 8080,
    protocol: "HTTP",
    targetType: "ip",
    vpcId: vpc.vpcId,
    deregistrationDelay: 5
}, {dependsOn: [lb]});

const targetGroups = (targetGroupCount === 1)
    ?  [
        {
            arn: tg.arn,
            weight: 1,
        },
    ] : [
        {
            arn: tg.arn,
            weight: 1,
        },
        {
            arn: tg2.arn,
            weight: 2,
        },
    ];

const listener = new aws.lb.Listener("listener", {
    loadBalancerArn: lb.arn,
    port: 80,
    defaultActions: [{
        type: "forward",
        forward: {
            targetGroups: targetGroups,
          },
    }]
}, {ignoreChanges: ["defaultActions[*].forward.targetGroups[*].weight"]});


export const lburl = lb.dnsName;
export const listenerId = listener.id;

```

Script:

```bash

#!/usr/bin/env bash

set -euo pipefail

export AWS_PROFILE=devsandbox

pulumi destroy --yes

pulumi config set targetGroupCount 2
pulumi up --yes --skip-preview
pulumi stack export --file tg2-state.json

pulumi config set targetGroupCount 1
pulumi up --yes --skip-preview

```

### Output of `pulumi about`

```
CLI          
Version      3.120.0
Go Version   go1.22.4
Go Compiler  gc

Plugins
KIND      NAME    VERSION
resource  aws     6.40.0
resource  awsx    2.12.1
resource  docker  4.5.4
resource  docker  3.6.1
language  nodejs  unknown

Host     
OS       darwin
Version  14.5
Arch     arm64

This project is written in nodejs: executable='/Users/anton/bin/node' version='v18.18.2'

Current Stack: anton-pulumi-corp/aws-4079/dev3

TYPE                                                 URN
pulumi:pulumi:Stack                                  urn:pulumi:dev3::aws-4079::pulumi:pulumi:Stack::aws-4079-dev3
pulumi:providers:awsx                                urn:pulumi:dev3::aws-4079::pulumi:providers:awsx::default_2_12_1
awsx:ec2:Vpc                                         urn:pulumi:dev3::aws-4079::awsx:ec2:Vpc::lb-vpc
pulumi:providers:aws                                 urn:pulumi:dev3::aws-4079::pulumi:providers:aws::default_6_37_1
aws:ec2/vpc:Vpc                                      urn:pulumi:dev3::aws-4079::awsx:ec2:Vpc$aws:ec2/vpc:Vpc::lb-vpc
aws:ec2/internetGateway:InternetGateway              urn:pulumi:dev3::aws-4079::awsx:ec2:Vpc$aws:ec2/vpc:Vpc$aws:ec2/internetGateway:InternetGateway::lb-vpc
aws:ec2/subnet:Subnet                                urn:pulumi:dev3::aws-4079::awsx:ec2:Vpc$aws:ec2/vpc:Vpc$aws:ec2/subnet:Subnet::lb-vpc-ecs-public-subnet-1
aws:ec2/subnet:Subnet                                urn:pulumi:dev3::aws-4079::awsx:ec2:Vpc$aws:ec2/vpc:Vpc$aws:ec2/subnet:Subnet::lb-vpc-ecs-public-subnet-2
aws:ec2/routeTable:RouteTable                        urn:pulumi:dev3::aws-4079::awsx:ec2:Vpc$aws:ec2/vpc:Vpc$aws:ec2/subnet:Subnet$aws:ec2/routeTable:RouteTable::lb-vpc-ecs-public-subnet-1
aws:ec2/routeTable:RouteTable                        urn:pulumi:dev3::aws-4079::awsx:ec2:Vpc$aws:ec2/vpc:Vpc$aws:ec2/subnet:Subnet$aws:ec2/routeTable:RouteTable::lb-vpc-ecs-public-subnet-2
aws:ec2/routeTableAssociation:RouteTableAssociation  urn:pulumi:dev3::aws-4079::awsx:ec2:Vpc$aws:ec2/vpc:Vpc$aws:ec2/subnet:Subnet$aws:ec2/routeTable:RouteTable$aws:ec2/routeTableAssociation:RouteTableAssociation::lb-vpc-ecs-public-subnet-1
aws:ec2/route:Route                                  urn:pulumi:dev3::aws-4079::awsx:ec2:Vpc$aws:ec2/vpc:Vpc$aws:ec2/subnet:Subnet$aws:ec2/routeTable:RouteTable$aws:ec2/route:Route::lb-vpc-ecs-public-subnet-1
aws:ec2/routeTableAssociation:RouteTableAssociation  urn:pulumi:dev3::aws-4079::awsx:ec2:Vpc$aws:ec2/vpc:Vpc$aws:ec2/subnet:Subnet$aws:ec2/routeTable:RouteTable$aws:ec2/routeTableAssociation:RouteTableAssociation::lb-vpc-ecs-public-subnet-2
aws:ec2/route:Route                                  urn:pulumi:dev3::aws-4079::awsx:ec2:Vpc$aws:ec2/vpc:Vpc$aws:ec2/subnet:Subnet$aws:ec2/routeTable:RouteTable$aws:ec2/route:Route::lb-vpc-ecs-public-subnet-2
pulumi:providers:pulumi                              urn:pulumi:dev3::aws-4079::pulumi:providers:pulumi::default
pulumi:providers:aws                                 urn:pulumi:dev3::aws-4079::pulumi:providers:aws::default_6_40_0
aws:ec2/securityGroup:SecurityGroup                  urn:pulumi:dev3::aws-4079::aws:ec2/securityGroup:SecurityGroup::lbSg
aws:lb/loadBalancer:LoadBalancer                     urn:pulumi:dev3::aws-4079::aws:lb/loadBalancer:LoadBalancer::lb
aws:lb/targetGroup:TargetGroup                       urn:pulumi:dev3::aws-4079::aws:lb/targetGroup:TargetGroup::tg
aws:lb/targetGroup:TargetGroup                       urn:pulumi:dev3::aws-4079::aws:lb/targetGroup:TargetGroup::tg2
aws:lb/listener:Listener                             urn:pulumi:dev3::aws-4079::aws:lb/listener:Listener::listener


Found no pending operations associated with dev3

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/anton-pulumi-corp
User           anton-pulumi-corp
Organizations  anton-pulumi-corp, moolumi, pulumi
Token type     personal

Dependencies:
NAME            VERSION
@pulumi/awsx    2.12.1
@pulumi/pulumi  3.120.0
@types/node     18.19.34
typescript      5.4.5
@pulumi/aws     6.40.0

Pulumi locates its logs in /var/folders/gd/3ncjb1lj5ljgk8xl5ssn_gvc0000gn/T/com.apple.shortcuts.mac-helper// by default
```

### Additional context

N/A

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

--- a/changelog/pending/20240614--engine--fix-a-panic-when-ignoring-wildcard-values-with-arrays-of-different-length.yaml
+++ b/changelog/pending/20240614--engine--fix-a-panic-when-ignoring-wildcard-values-with-arrays-of-different-length.yaml
@@ -1,0 +1,4 @@
+changes:
+- type: fix
+  scope: engine
+  description: Fix a panic when ignoring wildcard values with arrays of different length

--- a/sdk/go/common/resource/properties_path.go
+++ b/sdk/go/common/resource/properties_path.go
@@ -459,6 +459,11 @@ func (p PropertyPath) reset(old, new PropertyValue, oldIsSecret, newIsSecret boo
 			} else if old.IsArray() && new.IsArray() {
 				oldArray := old.ArrayValue()
 				newArray := new.ArrayValue()
+				// If arrays are of different length then this is a path failure because we can't
+				// continue the search of this path down each PropertyValue.
+				if len(oldArray) != len(newArray) {
+					return false
+				}
 
 				for i := range oldArray {
 					if !p[1:].reset(oldArray[i], newArray[i], oldIsSecret, newIsSecret) {

--- a/sdk/go/common/resource/properties_path_test.go
+++ b/sdk/go/common/resource/properties_path_test.go
@@ -750,6 +750,30 @@ func TestReset(t *testing.T) {
 			nil,
 		},
 		{
+			"Nested wildcard, old array is longer fails",
+			PropertyPath{"root", "*", 0},
+			PropertyMap{"root": NewProperty([]PropertyValue{
+				NewProperty(1.0),
+				NewProperty(2.0),
+			})},
+			PropertyMap{"root": MakeSecret(NewProperty([]PropertyValue{
+				NewProperty(3.0),
+			}))},
+			nil,
+		},
+		{
+			"Nested wildcard, new array is longer fails",
+			PropertyPath{"root", "*", 0},
+			PropertyMap{"root": NewProperty([]PropertyValue{
+				NewProperty(1.0),
+			})},
+			PropertyMap{"root": MakeSecret(NewProperty([]PropertyValue{
+				NewProperty(3.0),
+				NewProperty(4.0),
+			}))},
+			nil,
+		},
+		{
 			"Nested path, secret old",
 			PropertyPath{"root", "secret"},
 			PropertyMap{"root": MakeSecret(NewProperty(PropertyMap{"secret": NewProperty(1.0)}))},

